Introduce a router_service

o0Ignition0o opened the pull request:
Fixes #1496 eventually.

todo:
- [x] fix tests
- [x] enable APQ
- [x] deal with sandbox and homepage
- [x] decompression
- [x] tracing
- [x] `impl TryFrom<supergraph::Request> for Request` should put the graphql::Request in the body if the method is POST, and urlencode it in the path if the method is GET
- [ ] sure could use a helper to go from a defer chunked payload into a stream of graphql::Response for test purposes (like take the bytes, strip --graphql and everything, and deserialize the json contents), possibly using https://github.com/scottlamb/multipart-stream-rs/blob/master/examples/client.rs#L35
- [ ] move back the supergraph_service defer tests 
- [ ] not happy with `from_supergraph_mock_callback_and_configuration` and the likes, let's maybe use a builder
- [ ] router::Request and router::Response builders
- [x] Update and verify Apollo telemetry

bnjjj commented on apollo-router/src/services/router_service.rs:
```diff
@@ -0,0 +1,1157 @@
+                        {
+                            parts
+                                .headers
+                                .insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
```
NIT, I think we should use a constant for `application/json`

bnjjj commented on apollo-router/src/services/router_service.rs:
```diff
@@ -0,0 +1,1157 @@
+
+impl<SF> ServiceFactory<router::Request> for RouterCreator<SF>
+where
+    SF: StuffThatHasPlugins + ServiceFactory<supergraph::Request> + Clone + Send + Sync + 'static,
```
Am I the only one to find this name `StuffThatHasPlugins` a little bit weird ? :p

bnjjj replied:
I suggest to rename it to `HasPlugins`

o0Ignition0o replied:
Haha fair, FWIW I forgot to rename it, let's go with HasPlugins

bnjjj commented on apollo-router/src/services/router_service.rs:
```diff
@@ -0,0 +1,1157 @@
+        Error = BoxError,
+        Future = BoxFuture<'static, router::ServiceResult>,
+    > + Send {
+        let router_service = content_negociation::RouterLayer {}
```
Would be nice to have a `new` function on `RouterLayer`

bnjjj commented on apollo-router/src/services/supergraph_service.rs:
```diff
@@ -186,7 +189,13 @@ where
             let is_deferred = plan.is_deferred(operation_name.as_deref(), &variables);
-            if is_deferred && !accepts_multipart(req.supergraph_request.headers()) {
+
+            let accepts_multipart: bool = context
+                .get("accepts-multipart")
+                .unwrap_or_default()
+                .unwrap_or_default();
```
constant + namespace

bnjjj commented on apollo-router/src/services/supergraph_service.rs:
```diff
@@ -340,55 +349,81 @@ impl PluggableSupergraphServiceBuilder {
-    fn new_service(&self) -> Self::Service {
-        self.make().boxed()
-    }
+pub(crate) trait StuffThatHasPlugins {
```
```suggestion
pub(crate) trait HasPlugins {
```

bnjjj commented on apollo-router/src/test_harness.rs:
```diff
@@ -199,50 +205,49 @@ impl<'a> TestHarness<'a> {
+        .boxed_clone())
+    }
+
+    /// Builds the GraphQL service
```
```suggestion
    /// Builds the Router service
```

bnjjj commented on apollo-router/src/test_harness.rs:
```diff
@@ -199,50 +205,49 @@ impl<'a> TestHarness<'a> {
+        Ok((config, supergraph_creator))
     }
 
     /// Builds the GraphQL service
```
```suggestion
    /// Builds the Supergraph service
```

o0Ignition0o commented on apollo-router/tests/integration_tests.rs:
```diff
@@ -926,17 +942,20 @@ async fn query_with_router(
-            .to_vec()
-            .as_slice(),
+    serde_json::from_str(
+        dbg!(std::str::from_utf8(
```
it helps when the test fails, but i can also remove it yeah

o0Ignition0o commented on apollo-router/src/router.rs:
```diff
@@ -36,49 +36,42 @@ use self::Event::UpdateSchema;
-        .map_response(|response| {
-            response.map(|body| {
+        .map_request(|req: router::Request| req.router_request)
+        .map_err(|error| match error {})
```
turns out what we need is `!` and not `()` so this cannot work

o0Ignition0o commented on apollo-router/src/services/router_service.rs:
```diff
@@ -0,0 +1,1157 @@
+                };
+
+                let SupergraphResponse { response, context } =
+                    supergraph_service.oneshot(request).await.unwrap();
```
i ll go for ? instead, nice catch!

o0Ignition0o commented on apollo-router/src/services/layers/content_negociation.rs:
```diff
@@ -0,0 +1,271 @@
+                let accepts_wildcard = accepts_wildcard(req.router_request.headers());
+
+                if accepts_wildcard || accepts_multipart || accepts_json {
+                    req.context
+                        .insert("accepts-wildcard", accepts_wildcard)
+                        .unwrap();
+                    req.context
+                        .insert("accepts-multipart", accepts_multipart)
+                        .unwrap();
+                    req.context.insert("accepts-json", accepts_json).unwrap();
```
wfm

Geal commented on apollo-router/tests/integration_tests.rs:
```diff
@@ -540,9 +589,22 @@ async fn normal_query_with_defer_accept_header() {
+    };
 
-    assert!(actual.errors.is_empty());
+    insta::assert_snapshot!(std::str::from_utf8(actual.to_vec().as_slice()).unwrap());
```
```suggestion
    insta::assert_snapshot!(serde_json::from_slice::<serde_json::Value>(&actual).unwrap());
```

This will conserve the snapshot shape for all integration tests

o0Ignition0o replied:
i m going to update this before the mob review

Geal replied:
Thanks, that will simplify the PR further

o0Ignition0o replied:
I lied, but i ll update it asap

o0Ignition0o replied:
fixed (in an elegant way i think :D) in 9c5f4908

Geal commented on apollo-router/src/spec/query.rs:
```diff
@@ -976,10 +976,6 @@ impl Query {
         response_path: Option<&Path>,
         path: &Path,
     ) -> bool {
-        println!(
```
woops, sorry :sweat:

o0Ignition0o replied:
I dont see it 👀

o0Ignition0o commented on apollo-router/src/router_factory.rs:
```diff
@@ -49,7 +48,7 @@ impl std::fmt::Debug for Endpoint {
 impl Endpoint {
     /// Creates an Endpoint given a path and a Boxed Service
-    pub fn new(path: String, handler: transport::BoxService) -> Self {
+    pub fn new(path: String, handler: router::BoxService) -> Self {
```
fixed in b9b95266

garypen commented on apollo-router-benchmarks/benches/basic_composition.rs:
```diff
@@ -10,7 +10,7 @@ fn from_elem(c: &mut Criterion) {
         let builder = setup();
 
-        let router = runtime.block_on(builder.build()).unwrap();
+        let router = runtime.block_on(builder.build_router()).unwrap();
```
Ok. This was all answered in the walkthrough.

garypen reviewed (commented):
Huge amount of work. Really well done!!

Did you check the diffs of the documentation for public changes? Could you please document the result in the PR so we can see what has been added?

BrynCooke commented:
Here's a list of API changes:
```
Removed items from the public API
=================================
(none)

Changed items in the public API
===============================
(none)

Added items to the public API
=============================
+pub fn apollo_router::plugin::Plugin::router_service(&self, service: apollo_router::services::router::BoxService) -> apollo_router::services::router::BoxService
+pub mod apollo_router::services::router
+pub enum apollo_router::services::router::ParseError
+pub enum variant apollo_router::services::router::ParseError::InvalidUri(http::uri::InvalidUri)
+pub enum variant apollo_router::services::router::ParseError::SerializationError(serde_json::error::Error)
+pub enum variant apollo_router::services::router::ParseError::UrlEncodeError(serde_urlencoded::ser::Error)
+impl core::fmt::Debug for apollo_router::services::router::ParseError
+pub fn apollo_router::services::router::ParseError::fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result
+impl core::fmt::Display for apollo_router::services::router::ParseError
+pub fn apollo_router::services::router::ParseError::fmt(&self, formatter: &mut core::fmt::Formatter<'_>) -> core::fmt::Result
+impl core::error::Error for apollo_router::services::router::ParseError
+impl !core::panic::unwind_safe::RefUnwindSafe for apollo_router::services::router::ParseError
+impl core::marker::Send for apollo_router::services::router::ParseError
+impl core::marker::Sync for apollo_router::services::router::ParseError
+impl core::marker::Unpin for apollo_router::services::router::ParseError
+impl !core::panic::unwind_safe::UnwindSafe for apollo_router::services::router::ParseError
+impl<T> core::any::Any for apollo_router::services::router::ParseError where T: 'static + core::marker::Sized
+impl<T> downcast::Any for apollo_router::services::router::ParseError where T: core::any::Any
+pub fn apollo_router::services::router::ParseError::into_any(self: alloc::boxed::Box<T, alloc::alloc::Global>) -> alloc::boxed::Box<(dyn core::any::Any + 'static), alloc::alloc::Global>
+pub fn apollo_router::services::router::ParseError::into_any_rc(self: alloc::rc::Rc<T>) -> alloc::rc::Rc<(dyn core::any::Any + 'static)>
+pub fn apollo_router::services::router::ParseError::type_id(&self) -> core::any::TypeId
+pub fn apollo_router::services::router::ParseError::type_name(&self) -> &'static str
+impl<T> downcast::AnySync for apollo_router::services::router::ParseError where T: downcast::Any + core::marker::Send + core::marker::Sync
+pub fn apollo_router::services::router::ParseError::into_any_arc(self: alloc::sync::Arc<T>) -> alloc::sync::Arc<(dyn core::any::Any + core::marker::Sync + core::marker::Send + 'static)>
+impl<T> core::borrow::Borrow<T> for apollo_router::services::router::ParseError where T: core::marker::Sized
+pub fn apollo_router::services::router::ParseError::borrow(&self) -> &T
+impl<T> core::borrow::BorrowMut<T> for apollo_router::services::router::ParseError where T: core::marker::Sized
+pub fn apollo_router::services::router::ParseError::borrow_mut(&mut self) -> &mut T
+impl<T> core::convert::From<T> for apollo_router::services::router::ParseError
+pub fn apollo_router::services::router::ParseError::from(t: T) -> T
+impl<T> opentelemetry_api::trace::context::FutureExt for apollo_router::services::router::ParseError
+impl<T> tracing::instrument::Instrument for apollo_router::services::router::ParseError
+impl<T> tracing_futures::Instrument for apollo_router::services::router::ParseError
+impl<T, U> core::convert::Into<U> for apollo_router::services::router::ParseError where U: core::convert::From<T>
+pub fn apollo_router::services::router::ParseError::into(self) -> U
+impl<T> tonic::request::IntoRequest<T> for apollo_router::services::router::ParseError
+pub fn apollo_router::services::router::ParseError::into_request(self) -> tonic::request::Request<T>
+impl<D> owo_colors::OwoColorize for apollo_router::services::router::ParseError
+impl<E> core::any::Provider for apollo_router::services::router::ParseError where E: core::error::Error + core::marker::Sized
+pub fn apollo_router::services::router::ParseError::provide<'a>(&'a self, demand: &mut core::any::Demand<'a>)
+impl<T> typenum::type_operators::Same<T> for apollo_router::services::router::ParseError
+pub type apollo_router::services::router::ParseError::Output = T
+impl<T> alloc::string::ToString for apollo_router::services::router::ParseError where T: core::fmt::Display + core::marker::Sized
+pub fn apollo_router::services::router::ParseError::to_string(&self) -> alloc::string::String
+impl<T, U> core::convert::TryFrom<U> for apollo_router::services::router::ParseError where U: core::convert::Into<T>
+pub type apollo_router::services::router::ParseError::Error = core::convert::Infallible
+pub fn apollo_router::services::router::ParseError::try_from(value: U) -> core::result::Result<T, <T as core::convert::TryFrom<U>>::Error>
+impl<T, U> core::convert::TryInto<U> for apollo_router::services::router::ParseError where U: core::convert::TryFrom<T>
+pub type apollo_router::services::router::ParseError::Error = <U as core::convert::TryFrom<T>>::Error
+pub fn apollo_router::services::router::ParseError::try_into(self) -> core::result::Result<U, <U as core::convert::TryFrom<T>>::Error>
+impl<V, T> ppv_lite86::types::VZip<V> for apollo_router::services::router::ParseError where V: ppv_lite86::types::MultiLane<T>
+pub fn apollo_router::services::router::ParseError::vzip(self) -> V
+impl<T> tracing::instrument::WithSubscriber for apollo_router::services::router::ParseError
+impl<T> tracing_futures::WithSubscriber for apollo_router::services::router::ParseError
+#[non_exhaustive] pub struct apollo_router::services::router::Request
+pub struct field apollo_router::services::router::Request::context: apollo_router::Context
+pub struct field apollo_router::services::router::Request::router_request: http::request::Request<apollo_router::services::router::Body>
+impl core::convert::From<http::request::Request<hyper::body::body::Body>> for apollo_router::services::router::Request
+pub fn apollo_router::services::router::Request::from(router_request: http::request::Request<apollo_router::services::router::Body>) -> Self
+impl tower_service::Service<apollo_router::services::router::Request> for Handler
+pub type apollo_router::services::router::Request::Error = alloc::boxed::Box<(dyn core::error::Error + core::marker::Sync + core::marker::Send + 'static), alloc::alloc::Global>
+pub type apollo_router::services::router::Request::Future = tower::buffer::future::ResponseFuture<core::pin::Pin<alloc::boxed::Box<(dyn core::future::future::Future<Output = core::result::Result<<Handler as tower_service::Service<apollo_router::services::router::Request>>::Response, <Handler as tower_service::Service<apollo_router::services::router::Request>>::Error>> + core::marker::Send + 'static), alloc::alloc::Global>>>
+pub type apollo_router::services::router::Request::Response = apollo_router::services::router::Response
+pub fn apollo_router::services::router::Request::call(&mut self, req: apollo_router::services::router::Request) -> Self::Future
+pub fn apollo_router::services::router::Request::poll_ready(&mut self, cx: &mut core::task::wake::Context<'_>) -> core::task::poll::Poll<core::result::Result<(), Self::Error>>
+impl core::convert::TryFrom<apollo_router::services::supergraph::Request> for apollo_router::services::router::Request
+pub type apollo_router::services::router::Request::Error = apollo_router::services::router::ParseError
+pub fn apollo_router::services::router::Request::try_from(request: apollo_router::services::supergraph::Request) -> core::result::Result<Self, Self::Error>
+impl !core::panic::unwind_safe::RefUnwindSafe for apollo_router::services::router::Request
+impl core::marker::Send for apollo_router::services::router::Request
+impl core::marker::Sync for apollo_router::services::router::Request
+impl core::marker::Unpin for apollo_router::services::router::Request
+impl !core::panic::unwind_safe::UnwindSafe for apollo_router::services::router::Request
+impl<T> core::any::Any for apollo_router::services::router::Request where T: 'static + core::marker::Sized
+impl<T> downcast::Any for apollo_router::services::router::Request where T: core::any::Any
+pub fn apollo_router::services::router::Request::into_any(self: alloc::boxed::Box<T, alloc::alloc::Global>) -> alloc::boxed::Box<(dyn core::any::Any + 'static), alloc::alloc::Global>
+pub fn apollo_router::services::router::Request::into_any_rc(self: alloc::rc::Rc<T>) -> alloc::rc::Rc<(dyn core::any::Any + 'static)>
+pub fn apollo_router::services::router::Request::type_id(&self) -> core::any::TypeId
+pub fn apollo_router::services::router::Request::type_name(&self) -> &'static str
+impl<T> downcast::AnySync for apollo_router::services::router::Request where T: downcast::Any + core::marker::Send + core::marker::Sync
+pub fn apollo_router::services::router::Request::into_any_arc(self: alloc::sync::Arc<T>) -> alloc::sync::Arc<(dyn core::any::Any + core::marker::Sync + core::marker::Send + 'static)>
+impl<T> core::borrow::Borrow<T> for apollo_router::services::router::Request where T: core::marker::Sized
+pub fn apollo_router::services::router::Request::borrow(&self) -> &T
+impl<T> core::borrow::BorrowMut<T> for apollo_router::services::router::Request where T: core::marker::Sized
+pub fn apollo_router::services::router::Request::borrow_mut(&mut self) -> &mut T
+impl<T> core::convert::From<T> for apollo_router::services::router::Request
+pub fn apollo_router::services::router::Request::from(t: T) -> T
+impl<T> opentelemetry_api::trace::context::FutureExt for apollo_router::services::router::Request
+impl<T> tracing::instrument::Instrument for apollo_router::services::router::Request
+impl<T> tracing_futures::Instrument for apollo_router::services::router::Request
+impl<T, U> core::convert::Into<U> for apollo_router::services::router::Request where U: core::convert::From<T>
+pub fn apollo_router::services::router::Request::into(self) -> U
+impl<T> tonic::request::IntoRequest<T> for apollo_router::services::router::Request
+pub fn apollo_router::services::router::Request::into_request(self) -> tonic::request::Request<T>
+impl<D> owo_colors::OwoColorize for apollo_router::services::router::Request
+impl<T> typenum::type_operators::Same<T> for apollo_router::services::router::Request
+pub type apollo_router::services::router::Request::Output = T
+impl<T, U> core::convert::TryFrom<U> for apollo_router::services::router::Request where U: core::convert::Into<T>
+pub type apollo_router::services::router::Request::Error = core::convert::Infallible
+pub fn apollo_router::services::router::Request::try_from(value: U) -> core::result::Result<T, <T as core::convert::TryFrom<U>>::Error>
+impl<T, U> core::convert::TryInto<U> for apollo_router::services::router::Request where U: core::convert::TryFrom<T>
+pub type apollo_router::services::router::Request::Error = <U as core::convert::TryFrom<T>>::Error
+pub fn apollo_router::services::router::Request::try_into(self) -> core::result::Result<U, <U as core::convert::TryFrom<T>>::Error>
+impl<V, T> ppv_lite86::types::VZip<V> for apollo_router::services::router::Request where V: ppv_lite86::types::MultiLane<T>
+pub fn apollo_router::services::router::Request::vzip(self) -> V
+impl<T> tracing::instrument::WithSubscriber for apollo_router::services::router::Request
+impl<T> tracing_futures::WithSubscriber for apollo_router::services::router::Request
+#[non_exhaustive] pub struct apollo_router::services::router::Response
+pub struct field apollo_router::services::router::Response::context: apollo_router::Context
+pub struct field apollo_router::services::router::Response::response: http::response::Response<apollo_router::services::router::Body>
+impl apollo_router::services::router::Response
+pub async fn apollo_router::services::router::Response::into_graphql_response_stream(self) -> impl futures_core::stream::Stream<Item = core::result::Result<apollo_router::graphql::Response, serde_json::error::Error>>
+pub fn apollo_router::services::router::Response::map<F>(self, f: F) -> apollo_router::services::router::Response where F: core::ops::function::FnOnce(apollo_router::services::router::Body) -> apollo_router::services::router::Body
+pub async fn apollo_router::services::router::Response::next_response(&mut self) -> core::option::Option<core::result::Result<bytes::bytes::Bytes, hyper::error::Error>>
+impl core::convert::From<http::response::Response<hyper::body::body::Body>> for apollo_router::services::router::Response
+pub fn apollo_router::services::router::Response::from(response: http::response::Response<apollo_router::services::router::Body>) -> Self
+impl !core::panic::unwind_safe::RefUnwindSafe for apollo_router::services::router::Response
+impl core::marker::Send for apollo_router::services::router::Response
+impl core::marker::Sync for apollo_router::services::router::Response
+impl core::marker::Unpin for apollo_router::services::router::Response
+impl !core::panic::unwind_safe::UnwindSafe for apollo_router::services::router::Response
+impl<T> core::any::Any for apollo_router::services::router::Response where T: 'static + core::marker::Sized
+impl<T> downcast::Any for apollo_router::services::router::Response where T: core::any::Any
+pub fn apollo_router::services::router::Response::into_any(self: alloc::boxed::Box<T, alloc::alloc::Global>) -> alloc::boxed::Box<(dyn core::any::Any + 'static), alloc::alloc::Global>
+pub fn apollo_router::services::router::Response::into_any_rc(self: alloc::rc::Rc<T>) -> alloc::rc::Rc<(dyn core::any::Any + 'static)>
+pub fn apollo_router::services::router::Response::type_id(&self) -> core::any::TypeId
+pub fn apollo_router::services::router::Response::type_name(&self) -> &'static str
+impl<T> downcast::AnySync for apollo_router::services::router::Response where T: downcast::Any + core::marker::Send + core::marker::Sync
+pub fn apollo_router::services::router::Response::into_any_arc(self: alloc::sync::Arc<T>) -> alloc::sync::Arc<(dyn core::any::Any + core::marker::Sync + core::marker::Send + 'static)>
+impl<T> core::borrow::Borrow<T> for apollo_router::services::router::Response where T: core::marker::Sized
+pub fn apollo_router::services::router::Response::borrow(&self) -> &T
+impl<T> core::borrow::BorrowMut<T> for apollo_router::services::router::Response where T: core::marker::Sized
+pub fn apollo_router::services::router::Response::borrow_mut(&mut self) -> &mut T
+impl<T> core::convert::From<T> for apollo_router::services::router::Response
+pub fn apollo_router::services::router::Response::from(t: T) -> T
+impl<T> opentelemetry_api::trace::context::FutureExt for apollo_router::services::router::Response
+impl<T> tracing::instrument::Instrument for apollo_router::services::router::Response
+impl<T> tracing_futures::Instrument for apollo_router::services::router::Response
+impl<T, U> core::convert::Into<U> for apollo_router::services::router::Response where U: core::convert::From<T>
+pub fn apollo_router::services::router::Response::into(self) -> U
+impl<T> tonic::request::IntoRequest<T> for apollo_router::services::router::Response
+pub fn apollo_router::services::router::Response::into_request(self) -> tonic::request::Request<T>
+impl<D> owo_colors::OwoColorize for apollo_router::services::router::Response
+impl<T> typenum::type_operators::Same<T> for apollo_router::services::router::Response
+pub type apollo_router::services::router::Response::Output = T
+impl<T, U> core::convert::TryFrom<U> for apollo_router::services::router::Response where U: core::convert::Into<T>
+pub type apollo_router::services::router::Response::Error = core::convert::Infallible
+pub fn apollo_router::services::router::Response::try_from(value: U) -> core::result::Result<T, <T as core::convert::TryFrom<U>>::Error>
+impl<T, U> core::convert::TryInto<U> for apollo_router::services::router::Response where U: core::convert::TryFrom<T>
+pub type apollo_router::services::router::Response::Error = <U as core::convert::TryFrom<T>>::Error
+pub fn apollo_router::services::router::Response::try_into(self) -> core::result::Result<U, <U as core::convert::TryFrom<T>>::Error>
+impl<V, T> ppv_lite86::types::VZip<V> for apollo_router::services::router::Response where V: ppv_lite86::types::MultiLane<T>
+pub fn apollo_router::services::router::Response::vzip(self) -> V
+impl<T> tracing::instrument::WithSubscriber for apollo_router::services::router::Response
+impl<T> tracing_futures::WithSubscriber for apollo_router::services::router::Response
+pub type apollo_router::services::router::Body = hyper::body::body::Body
+pub type apollo_router::services::router::BoxCloneService = tower::util::boxed_clone::BoxCloneService<apollo_router::services::router::Request, apollo_router::services::router::Response, tower::BoxError>
+pub type apollo_router::services::router::BoxService = tower::util::boxed::sync::BoxService<apollo_router::services::router::Request, apollo_router::services::router::Response, tower::BoxError>
+pub type apollo_router::services::router::ServiceResult = core::result::Result<apollo_router::services::router::Response, tower::BoxError>
+impl core::fmt::Debug for apollo_router::services::supergraph::Request
+pub fn apollo_router::services::supergraph::Request::fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result
+impl core::convert::TryFrom<apollo_router::services::supergraph::Request> for apollo_router::services::router::Request
+pub type apollo_router::services::supergraph::Request::Error = apollo_router::services::router::ParseError
+pub fn apollo_router::services::supergraph::Request::try_from(request: apollo_router::services::supergraph::Request) -> core::result::Result<Self, Self::Error>
+pub fn apollo_router::Endpoint::from_router_service(path: alloc::string::String, handler: apollo_router::services::router::BoxService) -> Self
+pub async fn apollo_router::TestHarness::build_router(self) -> core::result::Result<apollo_router::services::router::BoxCloneService, tower::BoxError>
+pub async fn apollo_router::TestHarness::build_supergraph(self) -> core::result::Result<apollo_router::services::supergraph::BoxCloneService, tower::BoxError>
+pub fn apollo_router::TestHarness::router_hook(self, callback: impl core::ops::function::Fn(apollo_router::services::router::BoxService) -> apollo_router::services::router::BoxService + core::marker::Send + core::marker::Sync + 'static) -> Self
```

o0Ignition0o commented:
@BrynCooke thanks a lot for the diff

```
Removed items from the public API
=================================
(none)

Changed items in the public API
===============================
(none)
```

this looks good doesn't it? do we see things that went public and shouldnt ?

BrynCooke commented on apollo-router/src/services/router.rs:
```diff
@@ -0,0 +1,145 @@
+}
+
+impl Response {
+    pub async fn next_response(&mut self) -> Option<Result<Bytes, hyper::Error>> {
```
As we have a type alias for `Body` then also consider having one for `Error`

o0Ignition0o replied:
fixed in [64bf759](https://github.com/apollographql/router/pull/2170/commits/64bf7593f450e2359a03d76f8b432c2f4c28f13b)
